Objective-C/CBLDatabase.mm

```diff
@@ -790,13 +814,107 @@ - (void) freeC4DB {
 }
 
 
-#pragma mark - DOCUMENT SAVE AND CONFLICT RESOLUTION
+#pragma mark - DOCUMENT SAVE AND CONFLICT HANDLING
+
+
+- (BOOL) saveDocument: (CBLDocument*)document
+   concurrencyControl: (CBLConcurrencyControl)concurrencyControl
+           asDeletion: (BOOL)deletion
+                error: (NSError**)outError
+{
+    if (deletion && !document.revID) {
+        [NSException raise: NSInvalidArgumentException
+                    format: @"Do not allow to delete a newly created document "
+         "that has not been saved into the database."];
+        return NO;
+    }
+    
+    CBL_LOCK(self) {
+        if (![self prepareDocument: document error: outError])
+            return NO;
+        
+        // Begin a db transaction:
+        C4Transaction transaction(_c4db);
+        if (!transaction.begin())
+            return convertError(transaction.error(), outError);
+        
+        C4Document* curDoc = nil;
```

Comment:
I recommend using the utility class `c4::ref<T>` in `c4.hh`. It's a smart pointer that works with most C4 types, freeing them when it goes out of scope. Then you can yank out all the lines here that free `curDoc` and `newDoc`.

Reply:
I tried to use c4::ref. Giving that I'm not good at C++, I don't know how to make it retained when passing to keep the object in another class. Basically I would like to keep the C4Document (curDoc / newDoc) in the CBLC4Document class. Can you suggest me on this?

Reply:
Alternatively I could do @try @finally to free curDoc and newDoc in the finally block. What do you think?

Reply:
`@finally` is probably the simplest approach -- I hadn't thought of that.